Kernel/KLexicalPath.cpp

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2021, Max Wipfli <max.wipfli@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/KLexicalPath.h>
+
+namespace Kernel::KLexicalPath {
+
+bool is_canonical(StringView const& path)
+{
+    if (path.is_empty())
+        return false;
+    if (!path.starts_with('/'))
+        return false;
+    if (path.ends_with('/') && path.length() != 1)
+        return false;
+    if (path.contains("/./"sv) || path.ends_with("/."))
+        return false;
+    if (path.contains("/../"sv) || path.ends_with("/.."))
+        return false;
+    if (path.contains("//"sv))
+        return false;
+    return true;
+}
+
+StringView basename(StringView const& path)
+{
+    VERIFY(is_canonical(path));
+    auto slash_index = path.find_last('/');
+    if (!slash_index.has_value())
+        return path;
+    return path.substring_view(*slash_index + 1);
+}
+
+StringView dirname(StringView const& path)
+{
+    VERIFY(is_canonical(path));
+    auto slash_index = path.find_last('/');
+    // NOTE: In a Kernel context, you should never call dirname if there is only one path segment.
+    VERIFY(slash_index.has_value());
+    return path.substring_view(0, *slash_index);
+}
+
+Vector<StringView> parts(StringView const& path)
+{
+    VERIFY(is_canonical(path));
+    return path.split_view('/', true);
```

Comment:
`keep_empty` is meaningless, a canonical path cannot have empty segments.

Reply:
Same here.